File: fpga/test/tester.jl
include("../src/BAN.jl")
using .BAN
using Printf, Format

open("data.txt", "r") do ff
open("out_jl.txt", "w") do out

    SAMPLES =  parse(Int32, readuntil(ff," ")); #read(ff, Int32);
    SIZE = parse(Int32, readuntil(ff,"\n"));

    for i=1:SAMPLES
        p1 = parse(Int32, readuntil(ff," "));
        num1 = Vector{Float32}(undef, SIZE);
        for j=1:SIZE
            num1[j] = parse(Float32, readuntil(ff," "));
        end
        
        b1 = Ban(p1, num1);

        open("data.txt") do ff2
           
			for ii=1:(i+1)
				readline(ff2);
			end
		   
            for ii=(i+1):SAMPLES
                p2 = parse(Int32, readuntil(ff2," "));
                num2 = Vector{Float32}(undef, SIZE);
                for j=1:SIZE
                    num2[j] = parse(Float32, readuntil(ff2," "));
                end
                
                b2 = Ban(p2, num2);

                write(out, b1+b2);
				write(out, b1-b2);
				write(out, b1*b2);
				write(out, b1/b2);
				write(out, b2/b1);
				@printf(out, "%d\n", b1==b2);
				@printf(out, "%d\n", b1<b2);
				write(out, b1^2);
				write(out, b1^(-2));
				write(out, b1^5);
				write(out, b1^(-5));
				write(out, abs(b1));
				if b1>=0 && degree(b1)%2 == 0
					write(out, sqrt(b1));
				end
            end 
        end
    end
end
end

File: fpga/test/tester.cpp
//#include <iostream>
#include <fstream>
//#include <chrono>
#include "../src/ban.h"
using namespace std;
//using namespace std::chrono;

// Missing checks with reals (but 0)

int main(){

    ifstream ff;
    
    ofstream out;

    ff.open("data.txt");
    out.open("out_cpp.txt");

    unsigned SAMPLES, SIZE;
    ff>>SAMPLES>>SIZE;
    
    int p;
    T c[SIZE];

    //auto start = high_resolution_clock::now();

    for(unsigned i=0; i<SAMPLES; ++i){
        ff>>p;

        for(unsigned j=0; j<SIZE; ++j)
            ff>>c[j];

        Ban b1(p, c);
        ifstream ff2;
        ff2.open("data.txt");
        for(unsigned ii=0; ii <= i+1; ++ii)
            ff2.ignore(100, '\n');

        for(unsigned ii=i+1; ii<SAMPLES; ++ii){           
            ff2>>p;

            for(unsigned j=0; j<SIZE; ++j)
                ff2>>c[j];

            Ban b2(p, c);

            out<<b1+b2<<endl;
            out<<b1-b2<<endl;
            out<<b1*b2<<endl;
            out<<b1/b2<<endl;
            out<<b2/b1<<endl;
            out<<(b1==b2)<<endl;
            out<<(b1<b2)<<endl;
            out<<pow(b1,2)<<endl;
            out<<pow(b1,-2)<<endl;
            out<<pow(b1,5)<<endl;
            out<<pow(b1,-5)<<endl;
            out<<abs(b1)<<endl;
            if(b1>=0 && !(degree(b1) & 1u))
                out<<sqrt(b1)<<endl;
        }

        ff2.close();
    }

    /*
    auto stop = high_resolution_clock::now();

    auto duration = duration_cast<microseconds>(stop - start);
    cout << duration.count() << endl;
    */

    ff.close();
    out.close();

    return 0;
}